File: android/src/main/kotlin/com/ryanneilstroud/periscope_flutter/PeriscopeFlutterPlugin.kt
package com.ryanneilstroud.periscope_flutter

import android.os.Handler
import android.os.Looper
import com.ryanneilstroud.periscopeandroid.Periscope
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.OkHttpClient
import okhttp3.Request

/** PeriscopeFlutterPlugin */
class PeriscopeFlutterPlugin : FlutterPlugin, MethodCallHandler {
  private lateinit var channel: MethodChannel
  private val mainHandler = Handler(Looper.getMainLooper())

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "periscope_flutter")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "capture" -> capture(call, result)
      "stop" -> stop(result)
      "sendTestRequest" -> sendTestRequest(call, result)
      else -> result.notImplemented()
    }
  }

  private fun capture(call: MethodCall, result: Result) {
    val options = call.arguments as? Map<*, *>
    val receiver = (options?.get("receiver") as? Map<*, *>) ?: options
    val requestedPort = (receiver?.get("port") as? Number)?.toInt()
    val port = requestedPort ?: 61337

    if (port !in 1..65_535) {
      result.error("invalid_port", "Expected port in range 1...65535.", null)
      return
    }

    val host = (receiver?.get("host") as? String)?.trim()?.takeIf { it.isNotEmpty() }
    val periscopeReceiver = if (host == null) {
      Periscope.Receiver.simulator(port)
    } else {
      Periscope.Receiver.device(host, port)
    }

    Periscope.capture(periscopeReceiver)
    result.success(null)
  }

  private fun stop(result: Result) {
    Periscope.stop()
    result.success(null)
  }

  private fun sendTestRequest(call: MethodCall, result: Result) {
    val resolvedURLString = (call.arguments as? String)?.trim().takeUnless { it.isNullOrEmpty() }
      ?: "https://jsonplaceholder.typicode.com/todos/1"
    val url = resolvedURLString.toHttpUrlOrNull()

    if (url == null) {
      result.error("invalid_url", "sendTestRequest expected a valid URL string.", null)
      return
    }

    Thread {
      try {
        val client = OkHttpClient.Builder()
          .addInterceptor(Periscope.default.interceptor())
          .build()
        val request = Request.Builder()
          .url(url)
          .build()

        client.newCall(request).execute().use { response ->
          mainHandler.post { result.success(response.code) }
        }
      } catch (error: Exception) {
        mainHandler.post { result.error("request_failed", error.message, error.toString()) }
      }
    }.start()
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}

File: ios/Classes/PeriscopeFlutterPlugin.swift
import Flutter
import Foundation
import PeriscopeKit

public class SwiftPeriscopeFlutterPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "periscope_flutter", binaryMessenger: registrar.messenger())
    let instance = SwiftPeriscopeFlutterPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "capture":
      capture(call: call, result: result)
    case "stop":
      stop(result: result)
    case "sendTestRequest":
      sendTestRequest(call: call, result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func capture(call: FlutterMethodCall, result: @escaping FlutterResult) {
    let options = call.arguments as? [String: Any]
    let receiver = options?["receiver"] as? [String: Any] ?? options
    let requestedPort = receiver?["port"] as? NSNumber
    let portValue = requestedPort?.intValue ?? 61337

    guard (1...65_535).contains(portValue) else {
      result(FlutterError(code: "invalid_port", message: "Expected port in range 1...65535.", details: nil))
      return
    }

    let host = (receiver?["host"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
    let targetReceiver: Periscope.Receiver
    if let host, !host.isEmpty {
      targetReceiver = .device(host: host, port: portValue)
    } else {
      targetReceiver = .simulator(port: portValue)
    }

    Periscope.capture(for: targetReceiver)
    result(nil)
  }

  private func stop(result: @escaping FlutterResult) {
    Periscope.stop()
    result(nil)
  }

  private func sendTestRequest(call: FlutterMethodCall, result: @escaping FlutterResult) {
    let urlString = (call.arguments as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
    let resolvedURLString = (urlString?.isEmpty == false) ? urlString! : "https://jsonplaceholder.typicode.com/todos/1"

    guard let url = URL(string: resolvedURLString) else {
      result(FlutterError(code: "invalid_url", message: "sendTestRequest expected a valid URL string.", details: nil))
      return
    }

    let configuration = URLSessionConfiguration.default
    Periscope.inject(into: configuration)
    let session = URLSession(configuration: configuration)

    session.dataTask(with: url) { _, response, error in
      defer { session.finishTasksAndInvalidate() }

      if let error {
        DispatchQueue.main.async {
          result(FlutterError(code: "request_failed", message: error.localizedDescription, details: nil))
        }
        return
      }

      let statusCode = (response as? HTTPURLResponse)?.statusCode ?? 0
      DispatchQueue.main.async {
        result(statusCode)
      }
    }.resume()
  }
}
